src/EFCore.PG.NodaTime/Query/ExpressionTranslators/Internal/NpgsqlNodaTimeMemberTranslatorPlugin.cs

```diff
@@ -67,18 +67,53 @@ public NpgsqlNodaTimeMemberTranslator(ISqlExpressionFactory sqlExpressionFactory
                 return _sqlExpressionFactory.Constant(SystemClock.Instance);
 
             var declaringType = member.DeclaringType;
-            if (instance is not null
-                && (declaringType == typeof(LocalDateTime)
+            if (instance is not null)
+            {
+                if (declaringType == typeof(LocalDateTime)
                     || declaringType == typeof(LocalDate)
                     || declaringType == typeof(LocalTime)
-                    || declaringType == typeof(Period)))
-            {
-                return TranslateDateTime(instance, member, returnType);
+                    || declaringType == typeof(Period))
+                {
+                    return TranslateDateTime(instance, member, returnType);
+                }
+
+                if (declaringType == typeof(Duration))
+                {
+                    return TranslateDuration(instance, member);
+                }
             }
 
             return null;
         }
 
+        private SqlExpression TranslateDurationTotalMember(SqlExpression instance, double divisor) =>  _sqlExpressionFactory.Divide(GetDatePartExpressionDouble(instance, "epoch"), _sqlExpressionFactory.Constant(divisor));
+        
+        /// <summary>
+        /// Translates Duration members
+        /// </summary>
+        /// <param name="instance"></param>
+        /// <param name="member"></param>
+        /// <returns>The translated expression or null</returns>
+        private SqlExpression? TranslateDuration(SqlExpression instance, MemberInfo member)
+            => member.Name switch
+            {
+                nameof(Duration.TotalDays) => TranslateDurationTotalMember(instance, 86400),
+                nameof(Duration.TotalHours) => TranslateDurationTotalMember(instance, 3600),
+                nameof(Duration.TotalMinutes) => TranslateDurationTotalMember(instance, 60),
+                nameof(Duration.TotalSeconds) => TranslateDurationTotalMember(instance, 1),
+                nameof(Duration.TotalMilliseconds) => TranslateDurationTotalMember(instance, 0.001),
+                nameof(Duration.Days) => GetDatePartExpression(instance, "day"),
+                nameof(Duration.Hours) => GetDatePartExpression(instance, "hour"),
+                nameof(Duration.Minutes) => GetDatePartExpression(instance, "minute"),
+                nameof(Duration.Seconds) => GetDatePartExpression(instance, "second", true),
+                nameof(Duration.Milliseconds) => _sqlExpressionFactory.Convert(
+                    _sqlExpressionFactory.Multiply(
+                        _sqlExpressionFactory.Subtract(
+                            GetDatePartExpressionDouble(instance, "second", false), GetDatePartExpressionDouble(instance, "second", true)),
```

Comment:
This doesn't look right... Casting the end result from double to integer rounds, so if microseconds component is > 500, that will round up and add a millisecond. This can be corrected by adding a FLOOR (like how we get rid of the sub-second component to get seconds).

If we do this, we should also add the translation to TranslateDateTime below, as well as to the non-NodaTime BCL translator (NpgsqlDateTimeMemberTranslator, see test Where_datetime_millisecond_component in NorthwindWhereQueryNpgsqlTest). But as this translation is quite complicated/tricky and nobody has ever requested it, it's also OK to just leave it out of this PR and come back to it later upon request.

Reply:
I think you slightly misunderstand the code. We cast from double to int AFTER the multiplication.
this is needed for floating point rounding errors. If you omit the Convert, the unit test will fail since pgsql returns 19.999999 instead of 20.

the trick here is to have the "full seconds including ms" - "seconds without ms"  which leaves us with "subseconds". multiplying by 1000 gives the microseconds.

Reply:
Also: flooring said value would result in a wrong value. (19 instead of 20)

Reply:
I'm not saying to omit the convert - it's indeed needed to switch types (on the PG side) from double to int. However, I believe a FLOOR needs to be introduced right before the convert.

Assuming 18.200 as the starting value, the expression would be `CAST(FLOOR((18.200-18)*1000.0) AS INTEGER)`, which still yields 200 if I got my math right.

Now consider 18.200900 (PG has microsecond precision). As your code is currently written, we have `(18.200900-18)*1000.0`, which yields `200.9`. If you simply cast that to INTEGER, we get 201, which is incorrect, no?

Reply:
alright I see your point. but your solution doesn't work.

If we have 20 milliseconds, the stored value is 0.0199999995529651641845703125 if we floor it, it returns 19.

I guess the best thing is to exclude milliseconds for now, so no false values are returned.

Reply:
I agree... let's keep this focused on the other stuff - as I wrote above, we already don't translate milliseconds on all the other types, and nobody's complained...